### DISCUSSION TITLE
RavenDB with dynamic OID for databases problematic for SNMP Exporter

### DISCUSSION THREAD

**Discussion by mkocaks** [ravendb/ravendb#16573] [Q&A] [answered]
Hi,
After a few days of researching it seems it may not be possible for the following with snmp exporter for prometheus.
We are looking to monitor our RavenDB Cluster and all metrics via SNMP using SNMP Exporter with Prometheus, however the way RavenDB references the databases is problematic!
For each database the OID is different and we would rather not statically define in snmp.yaml each OID for each database (that can change in future if we remove and/or add databases to the raven cluster)
https://github.com/ravendb/ravendb/blob/v5.4/src/Raven.Server/Monitoring/Snmp/SnmpOids.cs

i.e,
[Description("Database name")]
public const string Name = "5.2.{0}.1.1

Is there a way we can dynamically walk through each unknown oid off the parent oid and get each database metrics with snmp exporter or is our only option to statically define these out of band and keep updating the snmp config each time we add or remove databases ?

I have also posted to the snmp exporter community
https://github.com/prometheus/snmp_exporter/discussions/884

**Comment by ppekrol** [comment]
Hi, the up to date list of OIDs is always available via `/monitoring/snmp/oids` endpoint. If needed you can parse it and update your file. My Q would be, if you really need to monitor each database separately? Most of the time the server-wide metrics are cover critical parts and maybe in your case it can suffice?

**Reply by mkocaks** [reply]
We would need to monitor each database as each database has different configurations (number of indexes, compex etc...), unless I am missing a key understanding ?
So as you said when we query this endpoint we would get different OID's for each database off of 1.3.6.1.4.1.45751.1.1.5.2.....
And rather than having to statically define the the next integer value for each custom DB (not the sys db) we ould rather have a way where this can be dynamically generated in snmp exporter, it seems to me Ravens implementation of SNMP is heavily focused around zabbix which can deal with such situations however does not help us when using other tools... (Prometheus is a Time Series Database, so once the label / metric is ingested if that OID changes for example when a database is removed or added then we risk corrupting the Prometheus TSDB as now the same OID references a different database name! )

**Reply by mkocaks** [reply]
If you could confirm that it is or not achievable this will help us to ditch or keep using the snmp license for raven

**Reply by mkocaks** [reply]
![image](https://github.com/ravendb/ravendb/assets/87523722/e3b2eba0-9c1f-4a6f-a0ce-9865a6cd77b4)

**Reply by ppekrol** [reply]
Internally we keep the mapping forever (database name => index), so even if you will delete the database and recreate it, it will get the same number again

**Reply by mkocaks** [reply]
Oh so if we we delete a database and then create a new one with a different name, this new one will NEVER get the previous freed up OID ? is this correct ?
But if we was to delete and recreate the database with the SAME name it will ALWAYS have the same OID ?

**Reply by ppekrol** [reply, marked as answer]
Correct. The information is stored in System database.

**Reply by mkocaks** [reply]
Thank you so its time to start statically defining out snmp config :-( but at least no risk of corrupting TSDB :)

**Reply by mkocaks** [reply]
Dear @ppekrol  We have just seen that a new database created has been given the OID integer value of 1 ?? ( We have many databases and would assume the very newest database created would be given the next number (i.e. 15) - why is it been given .1 ? 
Or do you randomly select the integer value for each databse name ?

**Reply by ppekrol** [reply]
Was SNMP enabled before those database were created? The 'index' is assigned when database is being created or when SNMP engine starts and a result of that mapping is the output of `/monitoring/snmp/oids` endpoint where all of those indexes are applied. So you should see unique indexes there per each database.

**Reply by mkocaks** [reply]
Nope it was enabled after :-) Thanks!
